src/lib/propagation/DAG.js

```diff
@@ -0,0 +1,63 @@
+import * as NodeFns from './nodeFns'
+import {SimulationNode} from './node'
+import * as constants from './constants'
+
+import * as _collections from 'gEngine/collections'
+
+export class SimulationDAG {
+  constructor(nodes) {
+    if (!!_.get(window, 'recorder')) { window.recorder.recordSimulationDAGConstructionStart(this) }
+    let rest = nodes.map(NodeFns.extractInputs)
+
+    let byId = _.transform(rest, (map, node) => {map[node.id] = {node, lastAncestors: node.inputs, ancestors: node.inputs}}, {})
+
+    const missingInputsNodes = _.remove(rest, NodeFns.hasMissingInputs(rest))
+    let graphErrorNodes = missingInputsNodes.map(NodeFns.withMissingInputError(nodes))
```

Comment:
graphErrorNodes contains more than just nodes with missing inputs; it also contains nodes with ancestral errors, or nodes with infinite loop errors. `errorNodes` is a superset of `graphErrorNodes`, containing additionally all nodes that come in with errors. These are stored separately as we extract all nodes with graph errors from the DAG prior to simulation, as they will never be able to be simulated successfully, whereas some nodes within `errorNodes` are re-simulated, in case a changed expression means the error will naturally be resolved. Also, `missingInputNodes` is used again, later, as the input to the `.map(withMissingInputErrorFn)` line below. I thought that would be a lot to put on one line. 

Unless I misunderstood your comments, I think those address your points here? Happy to make other changes, just don't want to make changes at the behest of incomplete information. I can also add more comments to this section to make things clearer.